### DISCUSSION TITLE
Trouble compiling with ThinLTO

### DISCUSSION THREAD

**Discussion by SaberToothKnaw** [google/filament#3750] [Q&A]
I'm trying to compile Filament and its samples on and for Ubuntu 20.04 with LTO enabled. I can compile the artifacts using Clang 7 with no issues, but when enabling LTO I get strange linker errors. Furthermore, when I edit the cmake-generated build.ninja file to use -flto=full instead of -flto=thin, compilation succeeds. Could I get a heads-up on if I'm doing something wrong or if this is a filament build system issue? Thanks.

Here are my steps:

1) in created directory [filament parent directory]/filament/out/cmake-release-lto , run cmake like so:
```
 cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=../release/filament -DFILAMENT_ENABLE_JAVA=OFF -DFILAMENT_ENABLE_LTO=ON -DFILAMENT_SUPPORTS_METAL=OFF -DFILAMENT_SUPPORTS_VULKAN=ON -DFILAMENT_INSTALL_BACKEND_TEST=OFF -DFILAMENT_USE_EXTERNAL_GLES3=OFF -DFILAMENT_USE_SWIFTSHADER=OFF -DFILAMENT_SKIP_SAMPLES=OFF -DCMAKE_C_COMPILER=clang-7 -DCMAKE_CXX_COMPILER=clang++-7 -DCMAKE_POLICY_DEFAULT_CMP0069=NEW ../..
```

2) run ninja:
```
ninja --verbose |& tee ./filament_thinlto_compile.txt
```
full ninja output log of the above:
[filament_thinlto_compile.txt](https://github.com/google/filament/files/6251783/filament_thinlto_compile.txt)
but in short, cntrl-f "error:", and around line 836 I get multiple linker errors like so:
> error: Failed to link module libs/image/libimage.a.llvm.5336.ImageOps.cpp: Expected at most one ThinLTO module per bitcode file

3) do a "ninja clean", and inside then generated "build.ninja" mass-replace all instances of "-flto=thin" with "-flto=full" with the cli or a text editor, do a ninja build again. For me, it works, and all the samples run fine.

Using LTO appears to have an effect, as build artifacts are a few hundred KBs smaller and linking takes forever, relative to non-LTO release builds with the same flags otherwise. I got the same results using Clang 10 instead of Clang 7 as well. Thanks in advance for comments.

**Comment by alexshpilkin** [comment]
Could this be a GNU ld bug, https://sourceware.org/bugzilla/show_bug.cgi?id=32153? It’s supposed to be fixed in binutils 2.44, but I’ve seen a very similar message in very similar circumstances with 2.44 on NixOS.
